src/main/java/com/google/api/codegen/configgen/DiscoInterfaceTransformer.java

```diff
@@ -0,0 +1,139 @@
+    Map<String, String> resourceNameMap = new TreeMap<>();
+    for (Method method : document.methods()) {
+      resourceNameMap.put(
+          methodToNamePatternMap.get(method),
+          getResourceIdentifier(method, apiInterface.getSimpleName(), resourceToNamePatternMap));
+    }
+    return resourceNameMap;
+  }
+
+  private Map<Method, String> getMethodToNamePatternMap(Document document) {
+    ImmutableMap.Builder<Method, String> methodToNamePatternMapBuilder = ImmutableMap.builder();
+    for (Method method : document.methods()) {
+      methodToNamePatternMapBuilder.put(method, DiscoGapicNamer.getCanonicalPath(method));
+    }
+
+    return methodToNamePatternMapBuilder.build();
+  }
+
+  @Override
+  public void generateResourceNameGenerations(ConfigNode parentNode, ApiModel model) {
+    Document document = ((DiscoApiModel) model).getDocument();
+    SetMultimap<String, String> resourceToNamePatternMap = getResourceToNamePatternMap(document);
+    FieldConfigNode resourceNameGenerationsNode =
+        MissingFieldTransformer.append("resource_name_generation", parentNode).generate();
+    if (NodeFinder.hasContent(resourceNameGenerationsNode.getChild())) {
+      return;
+    }
+
+    ConfigNode elementNode = new NullConfigNode();
+    ConfigNode prevNode = null;
+    for (Map.Entry<String, List<Method>> resource : document.resources().entrySet()) {
+      for (Method method : resource.getValue()) {
+        if (Strings.isNullOrEmpty(method.path())) {
+          continue;
+        }
+
+        int startLine = NodeFinder.getNextLine(prevNode == null ? parentNode : prevNode);
+        ConfigNode node = new ListItemConfigNode(startLine);
+        String messageName = DiscoGapicNamer.getRequestName(method).toUpperCamel();
+        ConfigNode messageNameNode =
+            FieldConfigNode.createStringPair(startLine, "message_name", messageName);
+        String parameterName =
+            DiscoGapicNamer.getResourceIdentifier(method.flatPath()).toLowerCamel();
+        String resourceName =
+            getResourceIdentifier(method, resource.getKey(), resourceToNamePatternMap);
+        ConfigNode fieldEntityMapNode =
+            new FieldConfigNode(NodeFinder.getNextLine(messageNameNode), "field_entity_map");
+        ConfigNode fieldEntityEntryNode =
+            FieldConfigNode.createStringPair(
+                NodeFinder.getNextLine(fieldEntityMapNode), parameterName, resourceName);
+        fieldEntityMapNode.setChild(fieldEntityEntryNode);
+        node.setChild(messageNameNode.insertNext(fieldEntityMapNode));
+
+        if (prevNode == null) {
+          resourceNameGenerationsNode.setChild(node);
+        } else {
+          prevNode.insertNext(node);
+        }
+
+        if (!elementNode.isPresent()) {
+          elementNode = node;
+        }
+
+        prevNode = node;
+      }
+    }
+  }
+
+  /** Map of base resource identifiers to all canonical name patterns that use that identifier. */
```

Comment:
I'm not sure what these words mean: "resource identifier", "canonical name pattern".

I would expect a resource identifier to refer to a particular resource's identifier (e.g., "projects/project-id/subscriptions/my-subscription-777"), which it would be surprising to encounter in a generator.

Reply:
This was reused from `DiscoConfigTransformer`. In this context:
- resource identifier: simple collection name
- canonical name pattern: the collection name pattern we use in GAPIC instead of the ones used in discovery